src/config_server.c

```diff
@@ -0,0 +1,2396 @@
+    } else if (hostkey->ks_type == KEYSTORE_REFERENCE) {
+        nc_server_del_keystore_reference(hostkey);
+    } else {
+        return;
+    }
+
+    nc_server_del_hostkey_name(hostkey);
+    opts->hostkey_count--;
+    if (!opts->hostkey_count) {
+        free(opts->hostkeys);
+        opts->hostkeys = NULL;
+    }
+}
+
+static void
+nc_server_del_auth_client_pubkey(struct nc_client_auth *auth_client, struct nc_client_auth_pubkey *pubkey)
+{
+    nc_server_del_auth_client_pubkey_name(pubkey);
+    nc_server_del_auth_client_pubkey_pub_base64(pubkey);
+
+    auth_client->pubkey_count--;
+    if (!auth_client->pubkey_count) {
+        free(auth_client->pubkeys);
+        auth_client->pubkeys = NULL;
+    }
+}
+
+static void
+nc_server_del_auth_client(struct nc_server_ssh_opts *opts, struct nc_client_auth *auth_client)
+{
+    uint16_t i, pubkey_count;
+
+    if (auth_client->ks_type == LOCAL_DEFINITON) {
+        pubkey_count = auth_client->pubkey_count;
+        for (i = 0; i < pubkey_count; i++) {
+            nc_server_del_auth_client_pubkey(auth_client, &auth_client->pubkeys[i]);
+        }
+    } else if (auth_client->ks_type == TRUSTSTORE_REFERENCE) {
+        nc_server_del_auth_client_ts_reference(auth_client);
+    } else {
+        return;
+    }
+
+    nc_server_del_auth_client_password(auth_client);
+    nc_server_del_auth_client_pam_name(auth_client);
+    nc_server_del_auth_client_pam_dir(auth_client);
+    nc_server_del_auth_client_username(auth_client);
+
+    opts->client_count--;
+    if (!opts->client_count) {
+        free(opts->auth_clients);
+        opts->auth_clients = NULL;
+    }
+}
+
+static void
+nc_server_del_ssh(struct nc_bind *bind, struct nc_server_ssh_opts *opts)
+{
+    uint16_t i, hostkey_count, client_count;
+
+    nc_server_del_local_address(bind);
+    if (bind->sock > -1) {
+        close(bind->sock);
+    }
+
+    /* store in variable because it gets decremented in the function call */
+    hostkey_count = opts->hostkey_count;
+    for (i = 0; i < hostkey_count; i++) {
+        nc_server_del_hostkey(opts, &opts->hostkeys[i]);
+    }
+
+    client_count = opts->client_count;
+    for (i = 0; i < client_count; i++) {
+        nc_server_del_auth_client(opts, &opts->auth_clients[i]);
+    }
+
+    nc_server_del_hostkey_algs(opts);
+    nc_server_del_kex_algs(opts);
+    nc_server_del_encryption_algs(opts);
+    nc_server_del_mac_algs(opts);
+
+    free(opts);
+    opts = NULL;
+}
+
+void
+nc_server_del_endpt_ssh(struct nc_endpt *endpt, struct nc_bind *bind)
+{
+    nc_server_del_endpt_name(endpt);
+    nc_server_del_ssh(bind, endpt->opts.ssh);
+
+    server_opts.endpt_count--;
+    if (!server_opts.endpt_count) {
+        free(server_opts.endpts);
+        free(server_opts.binds);
+        server_opts.endpts = NULL;
+        server_opts.binds = NULL;
+    }
+}
+
+/* presence container */
+int
+nc_server_configure_listen(const struct lyd_node *node, NC_OPERATION op)
+{
+    uint16_t i;
+
+    (void) node;
```

Comment:
I guess the parameter can then be removed, since it is not needed. I would also add assert for `op` to be either `NC_OP_DELETE` or `NC_OP_CREATE` to signify that both cases are handled.